xlators/storage/posix/src/posix-entry-ops.c

```diff
@@ -1184,6 +1200,47 @@ posix_unlink_gfid_handle_and_entry(call_frame_t *frame, xlator_t *this,
     return -1;
 }
 
+static int
+posix_unlink_stale_linkto(call_frame_t *frame, xlator_t *this,
+                          const char *real_path, int32_t *op_errno, loc_t *loc)
+{
+    int ret = -1;
+    struct iatt stbuf = {
+        0,
+    };
+
+    /* get the stale file gfid and stat-info */
+    ret = posix_pstat(this, NULL, NULL, real_path, &stbuf, _gf_false);
+    if (ret) {
+        if (errno == ENOENT) {
+            ret = 0; /* retry creation if file doesn't exist */
+            gf_msg(this->name, GF_LOG_INFO, errno, P_MSG_LSTAT_FAILED,
+                   "lstat on %s failed: file unlinked by another client",
+                   real_path);
+        } else {
+            gf_msg(this->name, GF_LOG_INFO, errno, P_MSG_LSTAT_FAILED,
+                   "lstat on %s failed", real_path);
+        }
+        goto out;
+    }
+    /* unlink only if linkto file*/
+    if (IS_DHT_LINKFILE_MODE(&stbuf)) {
+        gf_msg(this->name, GF_LOG_INFO, 0, P_MSG_HANDLE_CREATE,
+               "unlinking stale linkto: %s gfid: %s", real_path,
+               uuid_utoa(stbuf.ia_gfid));
+        ret = posix_unlink_gfid_handle_and_entry(
+            frame, this, real_path, &stbuf, op_errno, loc, _gf_false, NULL);
+
+    } else {
+        gf_msg(this->name, GF_LOG_DEBUG, 0, P_MSG_HANDLE_CREATE,
+               "skip unlinking stale data-file: %s gfid: %s", real_path,
+               uuid_utoa(stbuf.ia_gfid));
```

Comment:
We need to return some error in this case (`ret` is 0 here).

Reply:
It is being initialized: `int ret = -1;`

Reply:
Initialization value is overwritten at line 1213 and we only get here if `ret` was 0.